packages/block-editor/src/components/use-setting/index.js

```diff
@@ -93,126 +94,158 @@ const removeCustomPrefixes = ( path ) => {
 	return prefixedFlags[ path ] || path;
 };
 
+const mergeCache = new WeakMap();
+function mergeOrigins( value ) {
+	let result = mergeCache.get( value );
+	if ( ! result ) {
+		result = [ 'default', 'theme', 'custom' ].flatMap(
+			( key ) => value[ key ] ?? []
+		);
+		mergeCache.set( value, result );
+	}
+	return result;
+}
```

Comment:
it would be nice to have a docblock-style comment explaining the purpose of this function, and I wonder if the name could be clearer since `mergeOrigins` doesn't outwardly communicate that there's a cache involved.

I'm also confused on the value we're storing here, or why we need `flatMap()`. In the old code it's not clear to me either, so maybe this is why 🤷‍♂️ but it looks like it might be using `.concat()` with a default of `[]` to avoid needing a conditional for the case where `result` is missing the necessary key. I guess it's convenient for when keys are undefined and then it automatically returns an array with only those values.

but the old code was directly returning these things too while here we're storing them. since these are retrieved from `value` is there a need to split them out at this point? did you find that extracting these three properties was showing up in the hot path?

I'm so confused on all of this, really. is it worth adding a separate function and cache vs. directly building the array every time? maybe it was all confused by the nature of the original use of `reduce()`? I'm not trying to argue that the cache isn't worth it; I'm only asking if we have reason to believe it is.

```js
const values = [];
[ 'default', 'theme', 'custom' ].forEach(
	( key ) => {
		if ( value[ key ] ) {
			values.push( value[ key ] );
		}
	}
)

return values;
```

Reply:
Some settings, like `color.palette`, have a "raw" value where settings from different "origins" are in separate arrays:
```js
{
  default: [ ... ],
  theme: [ ... ],
  custom: [ ... ],
}
```
The `mergeOrigins` function merges the three arrays into one, concatenating them together. I added memoization so that two calls to `useSelect` with the same state and arguments return the same value. This was a little pre-existing bug.

I added a docblock to the function that contains this explanation.